pkg/3scale/amp/operator/upgrade.go

```diff
@@ -524,6 +532,214 @@ func (u *UpgradeApiManager) upgradeSystemPostgreSQLImageStream() (reconcile.Resu
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// deleteDeploymentConfigEnvVars deletes the environment variable names specified in
+// envVarNames from the given `desired` DeploymentConfig name. It deletes the
+// environment variables from all of its containers, initContainers, pre-hook
+// pods, post-hook pods and mid-hook pods.
+func (u *UpgradeApiManager) deleteDeploymentConfigEnvVars(desired *appsv1.DeploymentConfig, envVarNames []string) (reconcile.Result, error) {
+	existing := &appsv1.DeploymentConfig{}
+	err := u.Client().Get(context.TODO(), types.NamespacedName{Name: desired.Name, Namespace: u.apiManager.Namespace}, existing)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	update := false
+
+	// containers
+	for containerIdx := range existing.Spec.Template.Spec.Containers {
+		container := &existing.Spec.Template.Spec.Containers[containerIdx]
+		for _, envVarName := range envVarNames {
+			if envVarIdx := helper.FindEnvVar(container.Env, envVarName); envVarIdx >= 0 {
+				// remove index
+				container.Env = append(container.Env[:envVarIdx], container.Env[envVarIdx+1:]...)
+				update = true
+			}
+		}
+	}
+
+	// initContainers
+	for initContainerIdx := range existing.Spec.Template.Spec.InitContainers {
+		initContainer := &existing.Spec.Template.Spec.InitContainers[initContainerIdx]
+		for _, envVarName := range envVarNames {
+			if envVarIdx := helper.FindEnvVar(initContainer.Env, envVarName); envVarIdx >= 0 {
+				// remove index
+				initContainer.Env = append(initContainer.Env[:envVarIdx], initContainer.Env[envVarIdx+1:]...)
+				update = true
+			}
+		}
+	}
+
+	// pre-hook pod
+	var preHookPod *appsv1.ExecNewPodHook
+	if existing.Spec.Strategy.RollingParams != nil && existing.Spec.Strategy.RollingParams.Pre != nil && existing.Spec.Strategy.RollingParams.Pre.ExecNewPod != nil {
+		preHookPod = existing.Spec.Strategy.RollingParams.Pre.ExecNewPod
+	}
+	if existing.Spec.Strategy.RecreateParams != nil && existing.Spec.Strategy.RecreateParams.Pre != nil && existing.Spec.Strategy.RecreateParams.Pre.ExecNewPod != nil {
+		preHookPod = existing.Spec.Strategy.RecreateParams.Pre.ExecNewPod
+	}
+
+	if preHookPod != nil {
+		for _, envVarName := range envVarNames {
+			if envVarIdx := helper.FindEnvVar(preHookPod.Env, envVarName); envVarIdx >= 0 {
+				// remove index
+				preHookPod.Env = append(preHookPod.Env[:envVarIdx], preHookPod.Env[envVarIdx+1:]...)
+				update = true
+			}
+		}
+	}
+
+	// post-hook pod
+	var postHookPod *appsv1.ExecNewPodHook
+	if existing.Spec.Strategy.RollingParams != nil && existing.Spec.Strategy.RollingParams.Post != nil && existing.Spec.Strategy.RollingParams.Post.ExecNewPod != nil {
+		postHookPod = existing.Spec.Strategy.RollingParams.Post.ExecNewPod
+	}
+	if existing.Spec.Strategy.RecreateParams != nil && existing.Spec.Strategy.RecreateParams.Post != nil && existing.Spec.Strategy.RecreateParams.Post.ExecNewPod != nil {
```

Comment:
AFAIK `systrem-app`, `system-sphix` and `system-sidekiq` do not use `Spec.Strategy.RecreateParams `

Reply:
:+1: I was aware. The method was intended to be generic/reusable so I added that part too

Reply:
Is it being used by any other 3scale deploymentconfig? Not that I am aware of, But it could be. 

One more (small) thing, the code is replacing the `postHookPod` (or `preHookPod` earlier) in case there are recreateParams. I guess the strategy has to be either `rollingParams` (exclusive) OR `recreateParams`. Then, shouldn't it be `else if`?

Reply:
> Is it being used by any other 3scale deploymentconfig? Not that I am aware of, But it could be.
It isn't.

The strategy is exclusive. It could be `else if`. This should never happen though so having just another `if` is another issue. Even if it happens only one is used due to the type field so it would not be an issue